core/blockchain_pruner.go

```diff
@@ -0,0 +1,256 @@
+				for _, rangeStartEnd := range bp.deleteRange {
+					err := db.Compact(rangeStartEnd[0], rangeStartEnd[1])
+					if err != nil {
+						if isUnrecoverableErr(err) {
+							fmt.Printf("Unrecoverable error when writing leveldb: %v\nExitting\n", err)
+							os.Exit(1)
+						}
+					}
+				}
+				utils.Logger().Info().
+					Uint64("maxBlockNum", maxBlockNum).
+					Uint64("maxEpoch", maxEpoch.Uint64()).
+					Dur("cost", time.Now().Sub(startTime)).
+					Msg("pruneBeaconChain compact db success")
+			}
+		}()
+
+		bp.pruneBeaconChainBlock(db, maxBlockNum)
+		bp.pruneBeaconChainValidatorSnapshot(db, maxEpoch)
+	})
+}
+
+func (bp *blockchainPruner) pruneBeaconChainBlock(reader ethdb.Database, maxBlockNum uint64) {
+	minKey, maxKey := rawdb.IteratorBlocks(reader, func(blockNum uint64, hash common.Hash) bool {
+		if blockNum >= maxBlockNum {
+			return false
+		}
+
+		// skip genesis block
+		if blockNum == 0 {
+			return true
+		}
+
+		blockInfo := rawdb.ReadBlock(reader, hash, blockNum)
+		err := bp.deleteBlockInfo(blockInfo)
+		if err != nil {
+			utils.Logger().Info().
+				Uint64("blockNum", blockNum).
+				AnErr("err", err).
+				Msg("pruneBeaconChain delete block info error")
+			return false
+		}
+
+		err = rawdb.DeleteBlock(bp, hash, blockNum)
+		if err != nil {
+			utils.Logger().Info().
+				Uint64("blockNum", blockNum).
+				AnErr("err", err).
+				Msg("pruneBeaconChain delete block error")
+			return false
+		}
+
+		// limit time spent
+		bp.deletedBlockCount++
+		return bp.deletedBlockCount < maxDeleteBlockOnce
+	})
+
+	bp.deleteRange = append(bp.deleteRange, [2][]byte{minKey, maxKey})
+}
+
+func (bp *blockchainPruner) deleteBlockInfo(info *types.Block) error {
+	for _, cx := range info.IncomingReceipts() {
+		for _, cxReceipt := range cx.Receipts {
+			err := rawdb.DeleteCxLookupEntry(bp, cxReceipt.TxHash)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	for _, tx := range info.Transactions() {
+		err := rawdb.DeleteTxLookupEntry(bp, tx.Hash())
+		if err != nil {
+			return err
+		}
+	}
+
+	for _, stx := range info.StakingTransactions() {
+		err := rawdb.DeleteTxLookupEntry(bp, stx.Hash())
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (bp *blockchainPruner) pruneBeaconChainValidatorSnapshot(reader ethdb.Database, maxEpoch *big.Int) {
+	minKey, maxKey := rawdb.IteratorValidatorSnapshot(reader, func(addr common.Address, epoch *big.Int) bool {
+		if epoch.Cmp(maxEpoch) < 0 {
+			err := rawdb.DeleteValidatorSnapshot(bp, addr, epoch)
+			if err != nil {
+				utils.Logger().Info().
+					Str("addr", addr.Hex()).
+					Uint64("epoch", epoch.Uint64()).
+					AnErr("err", err).
+					Msg("pruneBeaconChain delete validator snapshot error")
+				return false
+			}
+
+			// limit time spent
+			bp.deletedValidatorSnapshot++
+		} else {
+			bp.skipValidatorSnapshot++
+		}
+
+		return bp.deletedValidatorSnapshot < maxDeleteBlockOnce
+	})
+
+	bp.deleteRange = append(bp.deleteRange, [2][]byte{minKey, maxKey})
```

Comment:
Could you specify a little more about how the deletion is worked?

Reply:
It's not `deleteRange`, It's `compactRange`, After levelDB deletes the data, the space cannot be released immediately, there is to release the deleted space quickly